src/item_location.cpp

```diff
@@ -310,113 +399,146 @@ class item_location::item_on_vehicle : public item_location::impl
                 cur.veh.remove_part( cur.part );
             } else {
-                item obj = cur.remove_item( *what );
-                if( obj.is_null() ) {
-                    debugmsg( "Tried to remove an item from a vehicle which doesn't contain it" );
-                }
+                cur.remove_item( *target() );
             }
-            what = nullptr;
         }
 };
 
 // use of std::unique_ptr<impl> forces these definitions within the implementation
-item_location::item_location() = default;
 item_location::item_location( item_location && ) = default;
 item_location &item_location::operator=( item_location && ) = default;
 item_location::~item_location() = default;
 
 const item_location item_location::nowhere;
 
+item_location::item_location()
+    : ptr( new impl::nowhere() ) {}
+
 item_location::item_location( const map_cursor &mc, item *which )
-    : ptr( new item_on_map( mc, which ) ) {}
+    : ptr( new impl::item_on_map( mc, which ) ) {}
 
 item_location::item_location( Character &ch, item *which )
-    : ptr( new item_on_person( ch, which ) ) {}
+    : ptr( new impl::item_on_person( ch, which ) ) {}
 
 item_location::item_location( const vehicle_cursor &vc, item *which )
-    : ptr( new item_on_vehicle( vc, which ) ) {}
+    : ptr( new impl::item_on_vehicle( vc, which ) ) {}
 
 bool item_location::operator==( const item_location &rhs ) const
 {
-    return ( ptr ? ptr->what : nullptr ) == ( rhs.ptr ? rhs.ptr->what : nullptr );
+    return ptr->target() == rhs.ptr->target();
 }
 
 bool item_location::operator!=( const item_location &rhs ) const
 {
-    return ( ptr ? ptr->what : nullptr ) != ( rhs.ptr ? rhs.ptr->what : nullptr );
+    return ptr->target() != rhs.ptr->target();
 }
 
 item_location::operator bool() const
 {
-    return ptr && ptr->what;
+    return ptr->valid();
 }
 
 item &item_location::operator*()
 {
-    return *ptr->what;
+    return *ptr->target();
 }
 
 const item &item_location::operator*() const
 {
-    return *ptr->what;
+    return *ptr->target();
 }
 
 item *item_location::operator->()
 {
-    return ptr->what;
+    return ptr->target();
 }
 
 const item *item_location::operator->() const
 {
-    return ptr->what;
+    return ptr->target();
+}
+
+void item_location::serialize( JsonOut &js ) const
+{
+    ptr->serialize( js );
+}
+
+void item_location::deserialize( JsonIn &js )
+{
+    auto obj = js.get_object();
+    auto type = obj.get_string( "type" );
+
+    int idx = -1;
+    tripoint pos = tripoint_min;
+
+    obj.read( "idx", idx );
+    obj.read( "pos", pos );
+
+    if( type == "character" ) {
+        ptr.reset( new impl::item_on_person( g->u, idx ) );
+
+    } else if( type == "map" ) {
+        ptr.reset( new impl::item_on_map( pos, idx ) );
+
+    } else if( type == "vehicle" ) {
+        auto *veh = g->m.veh_at( pos );
+        int part = obj.get_int( "part" );
+        if( veh && part <= int( veh->parts.size() ) ) {
```

Comment:
This comparison looks strange, it allows negative part values and values equal to `parts.size()`, both a are probably not valid. If they are valid, why is `parts.size() + 1` not valid as well?